### ISSUE TITLE
New Oscar architecture proposal

### ISSUE THREAD

**Issue by sasha0** [django-oscar/django-oscar#2972]
A while ago we discussed this with the @django-oscar/core team and finally I got to write it down as a proposal.

Since more and more projects implement only API and do not require frontend and/or dashboard, I came up with the raw idea of splitting Oscar into the parts, which would ease it customization:

- `oscar-core` - the main core functionality;
- `oscar-frontend` - basically all storefront views and templates;
- `oscar-dashboard` - all dashboard functionality.

On my latest projects we don't use storefront or dashboard at all (basically only core functionality and API), so in this case it won't need to disable almost all views.

Any feedback appreciated.

**Comment by maikhoepfel** [comment]
David and I always envisioned a split like that. Have a core with a API, and ship the dashboard and front end talking to that API. It would also make sure no business logic lands in the views, which is desirable. 
The front end and dashboard should be forbidden from making database calls and should only communicate via the API. But one could skip serialization - I don’t there’s a need to call the API over REST, I’d be fine with the API returning regular Django model instances.

**Comment by maerteijn** [comment]
I would be a proponent to split-up Oscar in different packages as @sasha0 suggested. From there you can go ahead and build a programmers API when there is a need for it.

Starting with separate packages / modules so it's easy to disable the store front end or dashboard (or not to install them at all) would really make sense, next to the fact that the maintenance and development of the codebase of the different components would be much easier.

**Comment by maikhoepfel** [comment]
What do you see as benefit of having seperate packages, @*maerteijn <https://github.com/maerteijn>*? I absolutely understand your point of wanting to disable them, but I think there's a hidden cost to splitting it up into packages - if you're referring to PyPi packages and seperate GitHub repos, not just Python packages.

If they were to ship as dedicated apps (the dashboard already is), one could easily disable it by not including it in INSTALLED_APPS. The frontend would require splitting up the current apps (this needs to be done either way), but I'm not sure there's a benefit to going for seperate packages. I'd be afraid that it makes maintenance and develeopment trickier, and it would also make releases trickier (because you need to make sure people install the correct versions of frontend - dashboard - core).

**Comment by sasha0** [comment]
> The frontend would require splitting up the current apps (this needs to be done either way), but I'm not sure there's a benefit to going for seperate packages.

@mvantellingen implemented POC of frontend disabling last year - https://github.com/django-oscar/django-oscar/commit/b4b7f6bbe8a9dac7784a2e8b526e040f0d6f8ec5, which demonstrates that it's not only to do with apps splitting in order to disable storefront.

>  I'd be afraid that it makes maintenance and develeopment trickier, and it would also make releases trickier (because you need to make sure people install the correct versions of frontend - dashboard - core).

In fact `oscar-storefront` and `oscar-dashboard` would be dependent on `oscar-core` and not interdependent, so there's not much complexity added.

**Comment by craigloftus** [comment]
I agree with the premise of separating the frontend, backend and core, but I also share @maikhoepfel's maintenance concerns with moving to separate packages.

It seems to me that it would only exacerbate the issues we already have maintaining feature 'parity' between the frontend, backend, core and rest api.

@sasha0 What are the advantages of using separate packages?

**Comment by maerteijn** [comment]
I agree on the part that maintenance would be more complex if you would split it up in multiple packages. You will need a version scheme guideline for the different packages (all `2.1.x` versions are compatible with each other as a simple example) to manage this properly.

On the other hand it makes maintenance easier to split the code into several packages as you would follow a real _separation of concerns_ approach: Code for the storefront is in the storefront package, has it's own issue-tracker, it's own roadmap and feature wish list and (hopefully) a dedicated maintainer or maintenance team.

The technical advantages of splitting it up inn multiple packages are the _dependencies_ and _deployment_ improvements. If we would have the following structure: 

```
                          +------------------+
                 +------->| oscar-storefront |
                 |        +------------------+
                 |
+------------+   |        +------------------+
| oscar-core |---+------->| oscar-dashboard  |
+------------+   |        +------------------+
                 |
                 |        +------------------+
                 +------->| oscar-api        |
                          +------------------+

```

And the dependencies per package (simplified example) are defined as:
```
oscar-core:
- django

oscar-storefront:
- oscar-core
- django-extra-views
- Pillow
- django-widget-tweaks

oscar-api:
- oscar-core
- restframework
```
And my webshop does not need the _storefront_ I also don't have to deploy `django-extra-views`, `Pillow`, `django-widget-tweaks` and other dependencies I don't use, including the _storefront_ code itself. 

Or another example: I could have 10 django instances with only _oscar-core_ and the _store-front_, and just 1 instance with  _oscar-core_  and  _oscar-dashboard_ installed.

I understand that all of this is a lot of work, so you could start with  several  (python) modules in the same repository before having own repositories.

**Comment by sasha0** [comment]
@maerteijn can't agree more - thanks for the thorough write up.

> Or another example: I could have 10 django instances with only oscar-core and the store-front, and just 1 instance with oscar-core and oscar-dashboard installed.

Same for us. In fact, the dashboard also get less relevant these days compared to SPA.

Also split up would potentially allow to have multiple storefronts - there are ready made ones https://github.com/moovweb/react-storefront, https://github.com/DivanteLtd/vue-storefront or etc that could be adopted.

**Comment by sasha0** [comment]
Also, looking at the history of changes of the previous major Oscar releases new changes affects mostly core functionality. 

Overall, we can firstly work on the further milestones with @django-oscar/core and afterwards it would be more visible about the plans regarding storefront and dashboard.

At the meantime, any input regarding architecture proposal appreciated.

**Comment by solarissmoke** [comment]
I think @maerteijn's proposed outline makes sense. What I don't know is what the best way is for these different components to communicate with each other. I feel that using a REST API is going to become unwieldy very quickly. I think it is worth exploring how other similar systems communicate, and what other mechanisms are available (RPC, perhaps?) before making what will be a significant change.

In the mean time I think the simpler step of just separating modules with the existing repo is a good place to start, and will help to identify issues that need to be addressed before a bigger split happens.

**Comment by SalahAdDin** [comment]
What's about GraphQL?

**Comment by pjstevns** [comment]
I'm all for @maerteijn's proposal. 
As for GraphQL: sounds like something that might go into oscar-api

**Comment by sasha0** [comment]
> As for GraphQL: sounds like something that might go into oscar-api

Or perhaps, `oscar-graphql`.

**Comment by maerteijn** [comment]
Exactly. REST / GraphGL / RPC are suitable for external applications or front-ends so they can use the "Oscar E-Commerce engine". 

If we have a storefront and/or dashboard written in Django it would not really make sense to "APIfy" this communication in such a level. Using a well thought Python API in the oscar-core package would be however very useful.

**Comment by saitam1** [comment]
This would also be an interesting structure

```
                                            +------------------+
                                   +------->| oscar-storefront |
                                   |        +------------------+
                                   |
+------------+    +------------+   |        +------------------+
| oscar-core |--->| oscar-api  |---+------->| oscar-dashboard  |
+------------+    +------------+   |        +------------------+
                                   |
                                   |        +------------------+
                                   +------->| api-endpoints    |
                                            +------------------+
```

**Comment by sasha0** [comment]
@saitam1 could you please elaborate more on this?

**Comment by saitam1** [comment]
I think this would be a more modern and modular approach. For the storefront, it would open up doors to connect frameworks like vue or react in a more easy way. 

Actually the dashboard could still be used without the api, because in most cases the dashboard probably does not need such a strong front-end customization.

**Comment by sasha0** [comment]
Within the @django-oscar/core team we agreed to start from decoupling interrelated components in the Oscar and @solarissmoke already has ideas on how to proceed.

**Comment by pawelmarkowski** [comment]
Modular approach would be great feature. As you wrote above it would be the point that allows people to connect Oscar with modern UI technologies. Graphql is an interesting point, but in my opinion classic rest API (DRF) should be still available.

**Comment by solarissmoke** [comment]
As far as decoupling components goes, the two things I would like to start with are `CheckoutSessionMixin` and `OrderPlacementMixin`. 

Both of these implement a huge amount of business logic in the views, which is extremely problematic for anyone that wants to use that logic outside of a view context. I think most projects end up basically reimplementing this logic.

I'd like to move the business logic currently living in those views into separate code that doesn't rely on a request/view context. This will also mean that things like sending order confirmation emails are not tied to views.

More generally I think we need to go through all the views and identify where there is business logic taking place that should be separated out.

**Comment by specialunderwear** [comment]
I always use the oscarapi serializer for checkout/orderplacement. That ensures the code I’m using is maintained and serializers are great general purpose object creation devices. If I need to separate checkout into steps, I just use a formwizzard to collect the data I need for the serializers in steps.

**Comment by specialunderwear** [comment]
I am currently rewriting the oscarapi serializers so they can be used for object creation/updating. I’m going to design it so they are the main interface for creating oscar entities in importers and such. So restapi is for http layer integration and serializers are for python level integration. Which amounts to the same.

I am mentioning this just as an update, I don’t really think this is an appropriate interface for first class (dashboard and storefront) clients. Just for third party integration.

**Comment by Aiky30** [comment]
Is there any work started on the ideas highlighted in this thread i.e. any branches / forks that I can look at? I'm particularly interested in using Django Oscar with Django CMS, I've been put off in the past due to the need to potentially rewrite the logic in the views and url mapping to closely couple the two without using the cms apphook feature i.e. one set of CMS templates with plugins to add Oscars e-commerce logic.

Separated apps as highlighted here open up the possibility of creating custom plugins in the cms for a "custom storefront" reusing the core of Oscar and potentially using the "Oscar dashboard" via a cms apphook.

I don't see many issues with the separated package communication, new Django CMS packages have dependencies from the core (pre release V4) and they just import the code / models as you would with any other Python library, they are in the end dependant on the core! As others have mentioned above the dashboard won't depend on the storefront and vice versa if designed properly. The key is getting the right code and business logic in the core.

**Comment by alekmlynek** [comment]
I'd love to know this as well (If work has started on this or branches available for viewing). Currently facing the decision to move forward with an e-commerce platform and the idea of this split would make Oscar a strong candidate, as that is our direction with all of our options.

**Comment by solarissmoke** [comment]
We've been making progress towards this, but there's still some way to go. Recent changes include refactoring Oscar's communications into a separate app. However we still have quite a bit more to do to decouple views from functional logic - no branches available yet for that I'm afraid.
